### PR TITLE
Add bend_pitch function

### LINKED ISSUE #9: Add a `bend_pitch` function
Currently, the only MIDI voice messages we manage are `note on` and `note off`. It would be interesting for `cardio-crumble` to have a richer pool of messages to choose from. I suggest adding another voice message to the `Midi` module. You can have a look at [this summary of midi channel voice messages](https://cmtext.indiana.edu/MIDI/chapter3_channel_voice_messages.php) for ideas and specifications.

One idea would be to add a `bend_pitch` function sending a `Pitch Bend` message. The signature could look as follows:
```
val bend_pitch : bend:char -> timestamp:int32 -> Event.t
```
where `bend` represents the pitch bend one wants to send in 14-bit resolution.

This issue is only about adding this new function to the `Midi` module (as a first step). In another issue later, we can use such a function.

### PR DESCRIPTION
**Pull request by kamtoeddy**
Resolves #9 

@pitag-ha, @Engil this is my solution

### CODE REVIEW

File: bin/midi.ml
Comment (Engil) [inline bin/midi.ml:6]: Actually I'd suggest taking the bend parameter as an `int` rather than a `char`.
Because `char` is only a 8 bit value, and you need here to be able to express up to a 14 bits value.
Can you make that change?

File: bin/midi.ml
Comment (Engil) [inline bin/midi.ml:9]: This does not compile as `data1` and `data2` are passed as integers. They should be `char`, you need to convert them again. :-)

**Review by Engil — COMMENTED** [review]
Hello!
Thank you for your PR, this is looking like a nice start. :-)
I made a few comments.
I could not compile the project to check, it seems you have type errors: did you try to compile your changes locally before submitting your PR?
If you have any issues with the initial setup (as described in the README), let us know and we can help you.

File: bin/midi.ml
Comment (kamtoeddy) [reply bin/midi.ml:6]: Ok

Comment (kamtoeddy) [reply bin/midi.ml:6]: When I was testing the function in `utop` I noticed that `char_of_int 11709` would not work

File: bin/midi.ml
Comment (kamtoeddy) [reply bin/midi.ml:9]: Sorry, I did not try to compile the code before pushing but I tested the function in `utop` but forgot to convert `data1` and `data2` to char

About setting up my local environment
- I have Ocaml installed in WSL
- The Ocaml extension for VSCode only worked on the first day

File: bin/midi.ml
Comment (Engil) [reply bin/midi.ml]: This is normal as you can only convert to char integers that can be contained in 8 bits. (limit would be 256 when working with unsigned integers.) :-) 

If your bitwise operations are correct, `data1` and `data2` should both be at most 7 bits long, so you shouldn't have any issue!

File: bin/midi.ml
Comment (kamtoeddy) [reply bin/midi.ml]: just pushed the fix

Comment (kamtoeddy) [reply bin/midi.ml]: > All good then! I enabled the CI to run on your PR so you will get immediate feedback on it next time you push!

It seems you have to approve for the CI to run

**Comment by Engil** [comment]
Thanks for the fixes!

I tested it locally and it seems to be working just fine, thank you!

One final adjustment: could you move the `bend_pitch` function in both `midi.ml` and `midi.mli` to be right after the `message_on/off` function?

**Comment by Engil** [comment]
By the way!
I tested it by sending a random bend pitch (up to `16000`) in the middle of each steps in the stat engine.
The result can be heard [here](https://drive.google.com/file/d/1Vi_j2K_yiIcqMss_cb81qmLbQ-uJJLaD/view?usp=share_link) :-)

**Comment by kamtoeddy** [comment]
> Thanks for the fixes!
> 
> I tested it locally and it seems to be working just fine, thank you!
> 
> One final adjustment: could you move the `bend_pitch` function in both `midi.ml` and `midi.mli` to be right after the `message_on/off` function?

Ok let me do that

**Comment by kamtoeddy** [comment]
@Engil the code has been updated

**Comment by Engil** [comment]
> Please is there a rule/format I could follow when writing code in the different modules?

This is more of an unspoken rule of developing with OCaml (or other programming languages!).

When you are designing an _interface_, you may want to group similar features in similar spots of the codebases. All of the functions (message_on, message_off, and bend_pitch) returns the same type: as such it makes sense to group them physically within the codebase.
There is no hard rule! (at least in this codebase). 😊 

I am going to approve this pull request: I'll let @pitag-ha merge it if she is ok with it.
Thank you for your contribution!

**Review by Engil — APPROVED** [review]

**Comment by Engil** [comment]
Although, before merging, it seems there's a liting issue with the CI!
Can you make the suggested changes so your PR passes the CI?

**Comment by kamtoeddy** [comment]
> > Please is there a rule/format I could follow when writing code in the different modules?
> 
> This is more of an unspoken rule of developing with OCaml (or other programming languages!).
> 
> When you are designing an _interface_, you may want to group similar features in similar spots of the codebases. All of the functions (message_on, message_off, and bend_pitch) returns the same type: as such it makes sense to group them physically within the codebase. There is no hard rule! (at least in this codebase). 😊
> 
> I am going to approve this pull request: I'll let @pitag-ha merge it if she is ok with it. Thank you for your contribution!

Oh that makes sense. Thanks

**Comment by kamtoeddy** [comment]
> Although, before merging, it seems there's a linting issue with the CI! Can you make the suggested changes so your PR passes the CI?

Yes I was actually struggling with this yesterday but I think I understand why the CI failed
The code has been updated

Is there another issue I can work on? The issues that are left are either assigned to other developers or they depend on an issue that has already been assigned ( like #18 )

**Comment by Engil** [comment]
I'll find you another issue to work on, I'll merge this one once CI is green. 😊

**Comment by Engil** [comment]
There is still one linting error (it looks like a missing or extra line return between two function signatures). I think ocamlformat should be taking care of this?

As for the issues, I opened https://github.com/pitag-ha/cardio-crumble/issues/21 just now, would you be happy to work on this? 😊

**Comment by kamtoeddy** [comment]
> There is still one linting error (it looks like a missing or extra line return between two function signatures). I think ocamlformat should be taking care of this?
> 
> As for the issues, I opened #21 just now, would you be happy to work on this? 😊

Hi @Engil,

I was not using ocamlformat then. I think the issue has been fixed now

I noticed that the formater complains about the extra line between the `bend_pitch` and `write_output` but complains if I remove the line between `message_on` and `message_off`. Why is that?

**Comment by Engil** [comment]
All good, I just merged your changes.
Thank you again for contributing!

**Comment by Engil** [comment]
As for the weird ocamlformat behaviour: I am not too accustomed to it, maybe @pitag-ha knows?

**Comment by pitag-ha** [comment]
Also from my side: thanks a lot @kamtoeddy, this looks great!

**Pull request merged**
